Adds Lewis Tunstall interview to blog (#262)
huggingface/blog: lewis-tunstall-interview.md

---
title: "Machine Learning Experts - Lewis Tunstall"
thumbnail: /blog/assets/60_lewis_tunstall_interview/thumbnail.png
---

<h1>Machine Learning Experts - Lewis Tunstall</h1>

<div class="blog-metadata">
    <small>Published April 11, 2022.</small>
</div>

<div class="author-card">
    <a href="/britneymuller">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1645809068511-5ef0ce775e979253a010ef4c.jpeg?w=200&h=200&f=face" title="Gravatar">
        <div class="bfc">
            <code>britneymuller</code>
            <span class="fullname">Britney Muller</span>
        </div>
    </a>
</div>

## 🤗 Welcome to Machine Learning Experts - Lewis Tunstall

> Hey friends! Welcome to Machine Learning Experts. I'm your host, Britney Muller and today’s guest is Lewis Tunstall. Lewis is a Machine Learning Engineer at Hugging Face where he works on applying Transformers to automate business processes and solve MLOps challenges.

Reviewer: Hey friends! Welcome to Machine Learning Experts. I'm your host, Britney Muller and today’s guest is [Lewis Tunstall](https://twitter.com/_lewtun). Lewis is a Machine Learning Engineer at Hugging Face where he works on applying Transformers to automate business processes and solve MLOps challenges.

> Hey friends! Welcome to Machine Learning Experts. I'm your host, Britney Muller and today’s guest is Lewis Tunstall. Lewis is a Machine Learning Engineer at Hugging Face where he works on applying Transformers to automate business processes and solve MLOps challenges.

Reviewer: ```suggestion
Hey friends! Welcome to Machine Learning Experts. I'm your host, Britney Muller and today’s guest is [Lewis Tunstall](https://twitter.com/_lewtun). Lewis is a Machine Learning Engineer at Hugging Face where he works on applying Transformers to automate business processes and solve MLOps challenges.
```

Reviewer: To provide context on Lewis early in the post.

*Note: Transcription has been slightly modified/reformatted to deliver the highest-quality reading experience.*

### Britney: Welcome, Lewis! Thank you so much for taking time out of your busy schedule to chat with me today about your awesome work!

**Lewis:** Thanks, Britney. It’s a pleasure to be here.

### Britney: Curious if you can do a brief self-introduction and highlight what brought you to Hugging Face? 

**Lewis:** What brought me to Hugging Face was transformers. In 2018, I was working with transformers at a startup in Switzerland. My first project was a question answering task where you input some text and train a model to try and find the answer to a question within that text.

In those days the library was called: pytorch-pretrained-bert, it was a very focused code base with a couple of scripts and it was the first time I worked with transformers. I had no idea what was going on so I read the original [‘Attention Is All You Need’](https://arxiv.org/abs/1706.03762) paper but I couldn’t understand it. So I started looking around for other resources to learn from. 

> In the process, Hugging Face exploded with their library growing into many architectures and I got really excited about contributing to open-source software. So around 2019, I had this kinda crazy idea to write a book about transformers because I felt there was an information gap that was missing. So I partnered up with my friend, [Leandro](https://twitter.com/lvwerra) [von Werra] and we sent [Tom](https://twitter.com/Thom_Wolf) [Wolf] a cold email out of nowhere saying, “Hey we are going to write a book about transformers, are you interested?” and I was expecting no response. But to our great surprise, he responded “Yea, sure let’s have a chat.” and around 1.5 years later this is our book: [NLP with Transformers](https://transformersbook.com/).

Reviewer 2: would be fun if there's a screenshot of this here

This collaboration set the seeds for Leandro and I to eventually join Hugging Face. And I've been here now for around nine months. 

### Britney: That is incredible. How does it feel to have a copy of your book in your hands? 

**Lewis:** I have to say, I just became a parent about a year and a half ago and it feels kind of similar to my son being born. You're holding this thing that you created. 

It's quite an exciting feeling and so different to actually hold it (compared to reading a PDF). Confirms that it’s actually real and I didn't just dream about it.

### Britney: Exactly. Congratulations!

Want to briefly read one endorsement that I love about this book;

> “Complexity made simple. This is a rare and precious book about NLP, transformers, and the growing ecosystem around them, Hugging Face. Whether these are still buzzwords to you or you already have a solid grasp of it all, the authors will navigate you with humor, scientific rigor, and plenty of code examples into the deepest secrets of the coolest technology around. From “off-the-shelf pre-trained” to “from-scratch custom” models, and from performance to missing labels issues, the authors address practically every real-life struggle of an ML engineer and provide state-of-the-art solutions, making this book destined to dictate the standards in the field for years to come.” 

Reviewer: ```suggestion
“_Complexity made simple. This is a rare and precious book about NLP, transformers, and the growing ecosystem around them, Hugging Face. Whether these are still buzzwords to you or you already have a solid grasp of it all, the authors will navigate you with humor, scientific rigor, and plenty of code examples into the deepest secrets of the coolest technology around. From “off-the-shelf pre-trained” to “from-scratch custom” models, and from performance to missing labels issues, the authors address practically every real-life struggle of an ML engineer and provide state-of-the-art solutions, making this book destined to dictate the standards in the field for years to come._” 
```

—Luca Perrozi Ph.D., Data Science and Machine Learning Associate Manager at Accenture.

### Britney: Can you talk about the work you've done with the transformers library?

**Lewis:** One of the things that I experienced in my previous jobs before Hugging Face was there's this challenge in the industry when deploying these models into production; these models are really large in terms of the number of parameters and this adds a lot of complexity to the requirements you might have. 

So for example, if you're trying to build a chatbot you need this model to be very fast and responsive. And most of the time these models are a bit too slow if you just take an off-the-shelf model, train it, and then try to integrate it into your application. 

> So what I've been working on for the last few months on the transformers library is providing the functionality to explore these models into a format that lets you run this much more efficiently using tools that we have at Hugging Face, but also just general tools in the open-source ecosystem.

Reviewer 2: ```suggestion
So what I've been working on for the last few months on the transformers library is providing the functionality to export these models into a format that lets you run them much more efficiently using tools that we have at Hugging Face, but also just general tools in the open-source ecosystem.
```

In a way, the philosophy of the transformers library is like writing lots of code so that the users don't have to write that code.

In this particular example, what we're talking about is something called the ONNX format. It's a special format that is used in industry where you can basically have a model that's written in PyTorch but you can then convert it to TensorFlow or you can run it on some very dedicated hardware.

And if you actually look at what's needed to make this conversion happen in the transformers library, it's fairly gnarly. But we make it so that you only really have to run one line of code and the library will take care of you. 

> So the idea is that this particular feature lets machine learning engineers or even data scientists take their model, converted to this format, and then optimize it to get faster latency and higher throughput. 

Reviewer 2: ```suggestion
So the idea is that this particular feature lets machine learning engineers or even data scientists take their model, convert it to this format, and then optimize it to get faster latency and higher throughput. 
```

### Britney: That's very cool. Have there been, any standout applications of transformers?

**Lewis:** Yeah. I think there are a few. One is maybe emotional or personal, for example many of us when OpenAI released GPT-2, this very famous language model which can generate text.

OpenAI actually provided in their blog posts some examples of the essays that this model had created. And one of them was really funny. One was an essay about why we shouldn't recycle or why recycling is bad.

And the model wrote a compelling essay on why recycling was bad. Leandro and I were working at a startup at the time and I printed it out and stuck it right above the recycling bin in the office as a joke. And people were like, “Woah, who wrote this?” and I said, “An algorithm.”

I think there's something sort of strangely human, right? Where if we see generated text we get more surprised when it looks like something I (or another human) might have written versus other applications that have been happening like classifying text or more conventional tasks.

> ### Britney: That's incredible. I remember when they released those examples for GTP-2, and one of my favorites (that almost gave me this sense of, whew, we're not quite there yet) were some of the more inaccurate mentions like “underwater fires”.

Reviewer 3: ```suggestion
### Britney: That's incredible. I remember when they released those examples for GPT-2, and one of my favorites (that almost gave me this sense of, whew, we're not quite there yet) were some of the more inaccurate mentions like “underwater fires”.
```

**Lewis:** Exactly!

**Britney:** But, then something had happened with an oil spill that next year, where there were actually fires underwater! And I immediately thought about that text and thought, maybe AI is onto something already that we're not quite aware of?

> ### You and other experts at Hugging Face have been working hard on the Hugging Face Course. How did that come about & where is it headced? 

Reviewer: ```suggestion
### You and other experts at Hugging Face have been working hard on the Hugging Face Course. How did that come about & where is it headed? 
```

> **Lewis:** When I joined Hugging Face, [Sylvian](https://twitter.com/GuggerSylvain) and [Lysandre](https://twitter.com/LysandreJik), two of the core maintainers of the transformers library, were developing a course to basically bridge the gap between people who are more like software engineers who are curious about natural language processing but specifically curious about the transformers revolution that's been happening. So I worked with them and others in the open-source team to create a free course called the [Hugging Face Course](https://huggingface.co/course/chapter1/1). And this course is designed to really help people go from knowing kind of not so much about NLP all the way through to having the ability to train models on many different tasks.

Reviewer 3: (nit: we're hoping to switch the course from NLP to ML at some point in the future)

And, we've released two parts of this course and planning to release the third part this year. I'm really excited about the next part that we're developing right now where we're going to explore different modalities where transformers are really powerful. Most of the time we think of transformers for NLP, but likely there's been this explosion where transformers are being used in things like audio or in computer vision and we're going to be looking at these in detail. 

### Britney: What are some transformers applications that you're excited about? 

**Lewis:** Yeah. So one that's kind of fun is in the course we had an event last year where we got people in the community to use the course material to build applications.

And one of the participants in this event created a cover letter generator for jobs. So the idea is that when you apply for a job there's always this annoying thing you have to write a cover letter and it's always like a bit like you have to be witty. So this guy created a cover letter generator where you provide some information about yourself and then it generates it from that.

And he actually used that to apply to Hugging Face.

### Britney: No way?!

> **Lewis:** Yea, he's joining the Big Science team as an intern. So. I mean this is a super cool thing, right? When you learn something and then use that thing to apply which I thought was pretty awesome. 

Reviewer 2: lol I didn't know they were joining haha nice story

Author: lol love that! Such a great story.

And at the moment, it's hard to find the answer to that, not just on the Hub, but in general in the space of machine learning this is quite hard. You often have to read papers and then you have to take those models and test them yourself manually and that's very slow and inefficient.

So one thing that we've been working on is to develop a way that you can evaluate models and datasets directly through the Hub. We're still trying to experiment there with the direction. But I'm hoping that we have something cool to show later this year. 

And there's another side to this which is that a large part of the measuring progress in machine learning is through the use of benchmarks. These benchmarks are traditionally a set of datasets with some tasks but what's been maybe missing is that a lot of researchers speak to us and say, “Hey, I've got this cool idea for a benchmark, but I don't really want to implement all of the nitty-gritty infrastructure for the submissions, and the maintenance, and all those things.”

> And so we've been working with some really cool partners on hosting benchmarks on the hub directly. So that then people in the research community can use the tooling that we have and then simplify the evaluation of these models. 

Reviewer 2: ```suggestion
And so we've been working with some really cool partners on hosting benchmarks on the Hub directly. So that then people in the research community can use the tooling that we have and then simplify the evaluation of these models. 
```

### Britney: That is super interesting and powerful.

> **Lewis:** Yeah. Maybe one thing to mention is that the whole question of evaluation is a very subtle one. For example, we know from previous benchmarks that most of the time when we design a task for example there's a very famous one called SQuAD which tries to measure how good models are at question answering.

Reviewer 2: nit: I would remove some of the "Yeah" at beginning of the replies

> **Lewis:** Yeah. Maybe one thing to mention is that the whole question of evaluation is a very subtle one. For example, we know from previous benchmarks that most of the time when we design a task for example there's a very famous one called SQuAD which tries to measure how good models are at question answering.

Reviewer 2: This paragraph was unclear to me. Most of the time what? 

Maybe this?

```
Maybe one thing to mention is that the whole evaluation question is a very subtle one. We know from previous benchmarks, such as SQuAD, a  famous benchmark to measure how good models are at question answering, that many of these transformer models are good at taking shortcuts. 
```

Author: Love that, thank you!

It resonated strongly. And I just love it because the explanations are super clear and you can share it with your family and your friends and say, “Hey, if you want to know what I'm doing? This can give you a rough idea.”

It gives you a very interesting insight into the Deep Mind researchers and their backstory as well.

### Britney: I'm definitely going to give that a listen. [Update: It’s one of my new favorite podcasts. :) Thank you, Lewis!]

### Britney: What are some of your favorite Machine Learning papers?

**Lewis:** Depends on how we measure this, but there's [one paper that stands out to me, which is quite an old paper](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf). It’s by the creator of random forests, Leo Breiman. Random forests is a very famous classic machine learning technique that's useful for tabular data that you see in industry and I had to teach random forests at university a year ago.

> And I was like, okay, I'll read this paper from the 200s and see if I understand it. And it's a model of clarity. It's very short, and very clearly explains how the algorithm is implemented. You can basically just take this paper and implement the code very very easily. And that to me was a really nice example of how papers were written in medieval times. 

Reviewer 2: ```suggestion
And I was like, okay, I'll read this paper from the 2000s and see if I understand it. And it's a model of clarity. It's very short, and very clearly explains how the algorithm is implemented. You can basically just take this paper and implement the code very very easily. And that to me was a really nice example of how papers were written in medieval times. 
```

But I think when you find a way to derive your own meaning and discover what you do is actually interesting and meaningful that that's the best part. Life is very up and down, right? At least for me personally, the things that have always been very meaningful are generally in creating things. So, I used to be a musician, so that was a way of creating music for other people and there was great pleasure in doing that. And now I kind of, I guess, create code which is a form of creativity.

### Britney: Absolutely. I think that's beautiful, Lewis! Is there anything else you would like to share or mention before we sign off? 

**Lewis:** Maybe [buy my book](https://transformersbook.com/).  

### Britney: It is so good!

**Lewis:** [shows book featuring a parrot on the cover] Do you know the story about the parrot?  

### Britney: I don't think so.

**Lewis:** So when O’Reilly is telling you “We're going to get our illustrator now to design the cover,” it's a secret, right?

> They don't tell you what the logic is or you have no say in the matter. So, basically, the illustrator comes up with an idea and in one of the last chapters of the book we have a section where we basically train a GPT-2 like model on Python code, this was Tom's idea, and he decided to call it code parrot.

Reviewer 2: ```suggestion
They don't tell you what the logic is or you have no say in the matter. So, basically, the illustrator comes up with an idea and in one of the last chapters of the book we have a section where we basically train a GPT-2 like model on Python code, this was Thom's idea, and he decided to call it code parrot.
```

### Britney: I love it. Well, it looks absolutely amazing. A lot of these types of books tend to be quite dry and technical and this one reads almost like a novel mixed with great applicable technical information, which is beautiful.

**Lewis:** Thanks. Yeah, that’s one thing we realized afterward because it was the first time we were writing a book we thought we should be sort of serious, right? But if you sort of know me I'm like never really serious about anything. And in hindsight, we should have been even more silly in the book.

I had to control my humor in various places but maybe there'll be a second edition one day and then we can just inject it with memes.

### Britney: Please do, I look forward to that!

**Lewis:** In fact, there is one meme in the book. We tried to sneak this in past the Editor and have the DOGE dog inside the book and we use a special vision transformer to try and classify what this meme is.

### Britney: So glad you got that one in there. Well done! Look forward to many more in the next edition. Thank you so much for joining me today. I really appreciate it. Where can our listeners find you online?

> **Lewis:** I'm fairly active on Twitter. You can just find me my handle [@_lewtun](https://twitter.com/_lewtun). LinkedIn is a strange place and I'm not really on there very much. And of course, there's the [Hugging Face Forums](https://discuss.huggingface.co/) or [Discord](https://discuss.huggingface.co/t/join-the-hugging-face-discord/11263).

Reviewer 3: you should also link to their profile on hf.co!

Author: Good call! I'll add that. Thanks, Julien.
